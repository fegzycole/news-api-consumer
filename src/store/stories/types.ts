export const SET_STORIES = "SET_STORIES";
export const SET_ERROR = "SET_ERROR";
export const SET_FETCHING_STORIES = "SET_FETCHING_STORIES";

export enum StoryCategories {
  TopHeadlines = "Top Headlines",
  Business = "Business",
  Entertainment = "Entertainment",
  General = "General",
  Health = "Health",
  Science = "Science",
  Sports = "Sports",
  Technology = "Technology",
}

export interface StoryData {
  source: {
    id: null | number;
    name: string;
  };
  author: string;
  title: string;
  description: string;
  url: string;
  urlToImage: string;
  publishedAt: Date | string;
  content: string;
}

export interface StoriesState {
  stories: StoryData[];
  fetchingStories: boolean;
  error?: string;
}

export interface AddStories {
  stories: StoryData[];
  clearStories: boolean;
}

export interface SetStories {
  type: typeof SET_STORIES;
  payload: AddStories;
}

export interface SetError {
  type: typeof SET_ERROR;
  payload: string;
}

export interface SetFetchingStories {
  type: typeof SET_FETCHING_STORIES;
  payload: boolean;
}

export type MoviesActionTypes =
  | SetStories
  | SetError
  | SetFetchingStories;
