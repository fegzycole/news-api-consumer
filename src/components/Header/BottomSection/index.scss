@import "../../../styles/mixins";
@import "../../../styles/palette";

.Categories_Container {
  padding: 1.25rem;
  width: 100%;
  overflow-x: scroll;
}

.Categories {
  width: 800px;
  @include flex-styles(row, space-between);

  @include lgScreen {
    margin: 0 auto;
  }
}

.Category_Btn {
  @include text-styles(500, 14px, $primary-color);
}

.Selected {
  color: $secondary-color;
}
