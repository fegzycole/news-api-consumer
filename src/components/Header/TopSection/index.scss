@import "../../../styles/mixins";

.TopSection {
  padding: 1rem;
  @include lgScreen {
    padding: 1.5rem 3rem;
  }
}

.Logo {
  display: grid;

  img {
    justify-self: center;
  }
}

.Nav {
  width: 100%;

  @include flex-styles(row, flex-end);
}

.Nav_Item {
  width: 300px;
  margin-top: -20px;
}

.Btn {
  border: none;
  background: $white-color;

  &:hover {
    cursor: pointer;
  }
}

.Actions {
  @include flex-styles("row", space-between);
  width: 100%;
}

.Actions_Item {
  @include text-styles(500, 14px, $primary-color, 20px);
}

.Hide_Sm {
  visibility: hidden;

  @include lgScreen {
    visibility: visible;
  }
}

.Subscribe {
  @include text-styles(500, 14, $white-color);
  background: $secondary-color;
  padding: 4px 16px;
  border-radius: 14px;
  margin-top: -3px;
}
