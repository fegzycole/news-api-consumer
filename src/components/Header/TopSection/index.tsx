import SearchIcon from "../../../assets/Union.svg";
import Logo from "../../../assets/logo.svg";
import Hamburger from "../../Hamburger";

import "./index.scss";

const TopSection = () => {
  return (
    <div className="TopSection">
      <div className="Logo">
        <img src={Logo} alt="Anderson Post Icon" />
      </div>
      <Hamburger />
      <nav className="Nav">
        <ul className="Nav_Item">
          <li className="Actions">
            <button className="Actions_Item Btn Hide_Sm">NewsLetter</button>
            <button className="Actions_Item Btn Hide_Sm">Sign In</button>
            <button className="Subscribe Btn Hide_Sm">Subscribe</button>
            <button className="Search_Icon Btn">
              <img src={SearchIcon} alt="Search Icon" />
            </button>
          </li>
        </ul>
      </nav>
    </div>
  );
};

export default TopSection;
