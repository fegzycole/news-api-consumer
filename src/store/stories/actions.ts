import { Dispatch } from "react";
import axios from "axios";

import {
  StoryData,
  SetStories,
  SET_STORIES,
  SetError,
  SET_ERROR,
  SetFetchingStories,
  SET_FETCHING_STORIES,
  StoryCategories,
  AddStories,
} from "./types";
import { getResponseError, getRequestPath } from "./helpers";

export const setStories = (payload: AddStories): SetStories => {
  return {
    type: SET_STORIES,
    payload,
  };
};

export const setError = (payload: string): SetError => {
  return {
    type: SET_ERROR,
    payload,
  };
};

export const setFetchingStories = (payload: boolean): SetFetchingStories => {
  return {
    type: SET_FETCHING_STORIES,
    payload,
  };
};

export const getStories = (
  category: StoryCategories,
  pageNumber: number
): ((
  dispatch: Dispatch<{ type: string; payload?: unknown }>
) => Promise<void>) => {
  return async (dispatch) => {
    dispatch(setError(""));
    dispatch(setFetchingStories(true));

    const requestPath = getRequestPath(category, pageNumber);

    try {
      const request = await axios.get(requestPath);

      const responseData = request.data.articles as StoryData[];

      const stories: StoryData[] = responseData as StoryData[];

      const payload: AddStories = {
        stories,
        clearStories: pageNumber === 1,
      };

      return dispatch(setStories(payload));
    } catch (error: any) {
      const renderedError = getResponseError(error);
      return dispatch(setError(renderedError));
    }
  };
};
