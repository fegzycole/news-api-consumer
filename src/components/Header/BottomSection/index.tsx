import { v4 as uuidv4 } from "uuid";

import { HeaderProps } from "../index";
import { StoryCategories } from "../../../store/stories/types";
import "./index.scss";

const categories = Object.values(StoryCategories).map((val) => val);

const BottomSection: React.FC<HeaderProps> = ({
  selectedCategory,
  handleClick,
}) => {
  return (
    <nav className="Categories_Container">
      <ul className="Categories">
        {categories.map((category) => (
          <li className="Category" key={uuidv4()}>
            <button
              onClick={() => handleClick(category)}
              className={`Category_Btn ${
                category === selectedCategory && `Selected`
              }`}
            >
              {category}
            </button>
          </li>
        ))}
      </ul>
    </nav>
  );
};

export default BottomSection;
