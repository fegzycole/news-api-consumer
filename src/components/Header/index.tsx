import TopSection from "./TopSection";
import BottomSection from "./BottomSection";
import Divider from "../Divider";
import { StoryCategories } from "../../store/stories/types";

export interface HeaderProps {
  selectedCategory: StoryCategories;
  handleClick: (val: StoryCategories) => void;
}

const Header: React.FC<HeaderProps> = ({ selectedCategory, handleClick }) => {
  return (
    <>
      <TopSection />
      <Divider />
      <BottomSection
        selectedCategory={selectedCategory}
        handleClick={handleClick}
      ></BottomSection>
      <Divider />
    </>
  );
};

export default Header;
