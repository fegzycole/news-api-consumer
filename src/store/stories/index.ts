import * as Types from './types';
import storiesReducer from './reducer';
import * as Actions from './actions';

export { Types, storiesReducer, Actions };
