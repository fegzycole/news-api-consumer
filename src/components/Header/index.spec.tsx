import { render } from "@testing-library/react";
import { StoryCategories } from "../../store/stories/types";
import Header, { HeaderProps } from "./index";

const mockProps: HeaderProps = {
  selectedCategory: StoryCategories.Entertainment,
  handleClick: jest.fn(),
};

jest.mock("./TopSection", () => ({
  __esModule: true,
  default: () => {
    return <div className="TopSection" />;
  },
  TopSection: true,
}));

jest.mock("./BottomSection", () => ({
  __esModule: true,
  default: () => {
    return <div className="BottomSection" />;
  },
  BottomSection: true,
}));

jest.mock("../Divider", () => ({
  __esModule: true,
  default: () => {
    return <div className="Divider" />;
  },
  Divider: true,
}));

describe("<Header />", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should match the Header component snapshot", () => {
    const view = render(<Header {...mockProps} />);
    expect(view.container).toMatchSnapshot();
  });
});
