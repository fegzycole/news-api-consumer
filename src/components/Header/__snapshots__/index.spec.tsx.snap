// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should match the Header component snapshot 1`] = `
<div>
  <div
    class="TopSection"
  />
  <div
    class="Divider"
  />
  <div
    class="BottomSection"
  />
  <div
    class="Divider"
  />
</div>
`;
